src/sdam/monitor.rs

```diff
@@ -56,84 +86,109 @@ pub(super) fn monitor_server(
-    conn: &mut Connection,
-    mut server_type: ServerType,
-    server: &Weak<Server>,
-) -> Option<ServerType> {
-    // If the server has been dropped, terminate the monitoring thread.
-    let server = match server.upgrade() {
-        Some(server) => server,
-        None => return None,
-    };
-
-    // If the topology has been dropped, terminate the monitoring thread.
-    let topology = match server.topology() {
-        Some(topology) => topology,
-        None => return None,
-    };
-
-    // Send an isMaster to the server.
-    let server_description = check_server(conn, server_type, &server);
-    server_type = server_description.server_type;
-
-    update_topology(topology, server_description);
-
-    Some(server_type)
-}
+    /// Checks the the server by running an `isMaster` command. If an I/O error occurs, the
+    /// connection will replaced with a new one.
+    fn check_server_and_update_topology(&mut self) {
+        // If the server has been dropped, terminate the monitoring thread.
+        let server = match self.server.upgrade() {
+            Some(server) => server,
+            None => return,
+        };
+
+        // If the topology has been dropped, terminate the monitoring thread.
+        let topology = match server.topology() {
+            Some(topology) => topology,
+            None => return,
+        };
+
+        // Send an isMaster to the server.
+        let server_description = self.check_server();
+        self.server_type = server_description.server_type;
+
+        update_topology(topology, server_description);
+    }
 
-fn check_server(
-    conn: &mut Connection,
-    server_type: ServerType,
-    server: &Arc<Server>,
-) -> ServerDescription {
-    let address = conn.address().clone();
+    fn check_server(&mut self) -> ServerDescription {
+        let address = self.address.clone();
 
-    match is_master(conn) {
-        Ok(reply) => return ServerDescription::new(address, Some(Ok(reply))),
-        Err(e) => {
-            server.clear_connection_pool();
+        match self.is_master() {
+            Ok(reply) => return ServerDescription::new(address, Some(Ok(reply))),
+            Err(e) => {
+                self.clear_connection_pool();
 
-            if server_type == ServerType::Unknown {
-                return ServerDescription::new(address, Some(Err(e)));
+                if self.server_type == ServerType::Unknown {
+                    return ServerDescription::new(address, Some(Err(e)));
+                }
             }
         }
+
+        ServerDescription::new(address, Some(self.is_master()))
     }
 
-    ServerDescription::new(address, Some(is_master(conn)))
-}
+    fn is_master(&mut self) -> Result<IsMasterReply> {
+        let result = self.is_master_inner();
+
+        if result
+            .as_ref()
+            .err()
+            .map(|e| e.kind.is_network_error())
+            .unwrap_or(false)
+        {
+            self.connection = self.make_new_connection()?;
```

Comment:
Because connection establishment could fail, I think it would make sense to have `self.connection` be an `Option`. Then when we encounter a network error, we null out the connection and only assign back to it if we successfully re-establish. Otherwise, we may attempt to use a broken/old connection next time we check the server.

Reply:
That seems reasonable to me; I'll update the code to do that.